File: python/searchable/src/searchable/client/geo.py
"""Great-circle distance helper for geo facets (mirrors packages/client/src/geo.ts)."""

import math

# Mean Earth radius in km (IUGG), same constant used by the TS client.
_EARTH_RADIUS_KM = 6371


def haversine_distance_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance between two (lat, lon) points in km, via the haversine
    formula (docs/guides/facets.md#geo-facets). Accurate enough for facet-radius
    filtering at this project's "small corpus JSON tier" scale -- the ~0.5% error
    from treating Earth as a perfect sphere is negligible next to a typical filter
    radius.
    """
    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return _EARTH_RADIUS_KM * c

File: packages/searchable/src/geo.ts
/** Mean Earth radius in km (IUGG), same constant used by python/searchable/src/searchable/client/geo.py. */
const EARTH_RADIUS_KM = 6371;

/**
 * Great-circle distance between two (lat, lon) points in km, via the
 * haversine formula (docs/guides/facets.md#geo-facets). Accurate enough for
 * facet-radius filtering at this project's "small corpus JSON tier" scale
 * (docs/guides/indexing.md#what-to-simplify-at-this-scale) -- the ~0.5%
 * error from treating Earth as a perfect sphere is negligible next to a
 * typical filter radius.
 */
export function haversineDistanceKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}
